### ISSUE TITLE
Improve performance of writing metrics

### ISSUE THREAD

**Issue by tilsche** [tud-zih-energy/otf2xx#2]
things like this: 

```
writer_.write(otf2::event::metric(otf2::chrono::time_point(otf2::chrono::duration(tv.timestamp)), metric_, values_));
```
are unfortunately super inefficient, due to creating and destroying vectors.

In write itself:
```
std::vector<OTF2_Type> type_ids(num_members);                                                        │
std::vector<OTF2_MetricValue> values(num_members); 
```

And in the temporary metric event itself, a copy of the data:
```
std::vector<value_container> values_;
```

We must find a way to do this without any dynamic allocation:

- Template case with static size?
- Custom allocators?

**Comment by bmario** [comment]
Things to keep in mind:

* ```OTF2_MetricValue``` is an union and ```OTF2_Type``` is an enum, which determines the used entry of the union.
* While ```value_container``` is a wrapper type, which handles the conversion between typesafe given data and the OTF2 union+enum representation.

Besides that, an obvious low hanging fruit is a special signature for the case with only one ```metric_member```, just like in the python bindings.

Maybe we should refactor this to something like: 
* ```value_container``` gets called ```value_wrapper``` or ```metric_box```
*  a special container for those things is called something like ```metric_container```. (No plural s involved here)

Having one special class for the collection of `metric_member` values, should eliminate the need for several different vector allocations.
IIRC the `metric_members` referenced in the ```value_container``` is an actual reference, which is sort of a bummer, when we like to reuse the `value_container` for different `metric_classes`.

**Comment by tilsche** [comment]
I'm not sure we need to mess with wrapping `vector<value_container>`. I was thinking of making `type_id` / `values` members of the writer thus allowing them to be reused across calls (just resize them to the maximum encountered member count). And `values_` in would have to be a ref for events where the user promises faithfully to use it only as a temporary.

**Comment by bmario** [comment]
Fixed in cdbb0b772a950818b323c0122c11d22d24b7d401
